common/src/main/java/io/pravega/common/util/OrderedItemProcessor.java

```diff
@@ -0,0 +1,253 @@
+     * all the items added before it have been processed.
+     * * If an item before this one failed to process and this item is still in the queue (it has not yet been processed),
+     * the returned future will be cancelled with an IllegalStateException to prevent out-of-order executions.
+     * * This method guarantees ordered execution as long as its invocations are serial. That is, it only guarantees that
+     * the item has begun processing or an order assigned if it returned successfully.
+     *
+     * @param item The item to process.
+     * @return A CompletableFuture that, when completed, will indicate that the item has been processed. This will contain
+     * the result of the processing function applied to this item.
+     */
+    public CompletableFuture<ResultType> process(ItemType item) {
+        Preconditions.checkNotNull(item, "item");
+        CompletableFuture<ResultType> result = null;
+        synchronized (this.stateLock) {
+            Exceptions.checkNotClosed(this.closed, this);
+            if (hasCapacity() && this.pendingItems.isEmpty()) {
+                // Not at capacity. Reserve a spot now.
+                this.activeCount++;
+            } else {
+                // We are at capacity or have a backlog. Put the item in the queue and return its associated future.
+                result = new CompletableFuture<>();
+                this.pendingItems.add(new QueueItem(item, result));
+            }
+        }
+
+        if (result == null) {
+            // We are below max capacity, so simply process the item, without queuing up.
+            // It is possible, under some conditions, that we may get in here when the queue empties up (by two spots at
+            // the same time). In that case, we need to acquire the processing lock to ensure that we don't accidentally
+            // process this item before we finish processing the last item in the queue.
+            synchronized (this.processingLock) {
+                result = processInternal(item);
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Callback that is invoked when an item has completed execution.
+     *
+     * @param exception (Optional) An exception from the execution. If set, it indicates the item has not been
+     *                  processed successfully.
+     */
+    @VisibleForTesting
+    protected void executionComplete(Throwable exception) {
+        Collection<QueueItem> toFail = null;
+        val failEx = new AtomicReference<Throwable>();
+        synchronized (this.stateLock) {
+            // Release the spot occupied by this item's execution.
+            this.activeCount--;
+            if (exception != null && !this.closed) {
+                // Need to fail all future items and close to prevent new items from being processed.
+                failEx.set(new ProcessingException("A previous item failed to commit. Cannot process new items.", exception));
```

Comment:
Yes, not using AtomicReference is cleaner. However it doesn't like that I used in a lambda expression below (since i had to initialize it to either null or some value). I removed the lambda and the AtomicReference.